### PR TITLE
feat: apollo graphql link and server

### PR DESCRIPTION
**Pull request by viglucci**
Adds experimental `ApolloLink` and `ApolloServer` implementations to support GraphQL over RSocket with [Apollo GraphQL](https://www.apollographql.com/).

### Motivation:

Spring [recently added support](https://github.com/spring-projects/spring-graphql/issues/339) for GraphQL over RSocket, and we were interested to see if we could provide similar support via rsocket-js. This addition will provide a mechanism for consuming RSocket transport using GraphQL servers via JS clients (browser etc.).

Additionally, these additions provide an interesting alternative to existing Apollo GraphQL [subscription links](https://www.apollographql.com/docs/react/data/subscriptions/), such as [graphql-ws](https://www.npmjs.com/package/graphql-ws). Existing GraphQL subscription users could now benefit from a potentially increased number of available transport protocol options. I believe it would be beneficial for these advancements to be more broadly broadcasted and advocated for in the GraphQL community.

### Modifications:

- added rsocket-graphql-apollo-link package
- added rsocket-graphql-apollo-server package
- provided examples for using new packages

### Result:

See above.

### Related:

- https://github.com/apollographql/apollo-server/issues/6264
- https://community.apollographql.com/t/runhttpquery-alternative-for-non-http-frameworks-apolloserver-rsocket-integration/3125

### Open questions/items:

- Are we improperly assuming a route will be used/required on the server?
- Are we sending the proper metadata mime type on SETUP and request?
- Are errors being handled according to spec?
- Test Interop with Spring server/client.

### CODE REVIEW

File: packages/rsocket-graphql-apollo-server/src/ApolloServer.ts [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+          method: "POST",
+          options: () => {
+            return this.createGraphQLServerOptions(payload);
+          },
+          query,
+          request: new Request("/graphql", {
+            headers: new Headers(),
+            method: "POST",
+          }),
+        })
```
Comment (viglucci) [inline packages/rsocket-graphql-apollo-server/src/ApolloServer.ts:124]: From what I can tell, `runHttpQuery` seems to be the best entry point into Apollo's logic for handling a query. This feels strange for our use case, seeing as we are not HTTP, so I'll likely reach out to Apollo and see if there is a better way to process a query that is completely divorced from HTTP concepts.

File: packages/rsocket-tcp-server/src/TcpDuplexConnection.ts [hunk 2]
```diff
@@ -71,6 +71,10 @@ export class TcpDuplexConnection
     this.socket.off("error", this.handleError);
     this.socket.off("data", this.handleData);
 
+    // TODO: should this be destroy instead of end()?
+    //   apparently client can still send data after calling end()
+    //   and socket will never fully close until peer closes their end.
+    //   ex: the above is a problem when connected to a .Net TCP socket
     this.socket.end();
```
Comment (OlegDokuka) [inline packages/rsocket-tcp-server/src/TcpDuplexConnection.ts:78]: how about creating a separate issue instead of making this comment?

Comment (viglucci) [reply packages/rsocket-tcp-server/src/TcpDuplexConnection.ts:78]: Ah yeah, good catch. Meant to move this to an issue.

**Comment by viglucci** [comment]
Moving this PR out of draft is pending investigating Apollo's `executeOperation`, as suggested [here](https://community.apollographql.com/t/runhttpquery-alternative-for-non-http-frameworks-apolloserver-rsocket-integration/3125).

File: packages/rsocket-tcp-server/src/TcpDuplexConnection.ts (hunk 2 above)
Comment (viglucci) [reply packages/rsocket-tcp-server/src/TcpDuplexConnection.ts]: Issues for reference: https://github.com/rsocket/rsocket-js/issues/226

**Comment by viglucci** [comment]
I will fix DCO on merge.

**Comment by viglucci** [comment]
Open question: are we improperly assuming a route will be required on the server?

**Comment by viglucci** [comment]
Rebased to fix resolve DCO check issue.

**Pull request merged**